[tmva][sofie] ROperator_Comparision on-the-fly broadcasting

olia110 opened the pull request:
# This Pull request:

This pull request refactors the ROperator_Comparision to handle broadcasting more efficiently.
Added new tests to verify the correctness of the ROperator_Comparision.

Commit:
[tmva][sofie] Add new comparison operator tests

sanjibansg commented on tmva/sofie/inc/TMVA/ROperator_Comparision.hxx:
```diff
@@ -77,7 +77,7 @@ public:
    ROperator_Comparision(const std::string & nameX1, const std::string & nameX2, const std::string & nameY):
```
@lmoneta do we need the `std::vector<Dim> fDimShapeX1;
   std::vector<Dim> fDimShapeX2;` here? They seem to be only used in two `if` cases

sanjibansg reviewed (approved):
LGTM! thanks for this change!

lmoneta reviewed (dismissed):
Thank you for fixing also this operator to avoid broadcasting and for adding the relevant test.
Apart for some issue of some deleted code which should be still valid,  the other changes are for following and being consistent what has been done in the BinaryOperator:
https://github.com/root-project/root/pull/19742/files

guitargeek commented:
Thank you @olia110 for these developments! I see they have made it into ROOT `master` via another commit (https://github.com/root-project/root/commit/03bcecfe705), so dropping the changes for the comparison operator from this PR, and keep only the added tests.

guitargeek committed:
[tmva][sofie] Use multidirectional broadcasting in comparison operators

ONNX comparison operators (Equal, Less, Greater, ...) follow numpy-style
multidirectional broadcasting, like the arithmetic binary operators. Using
UnidirectionalBroadcastShape fails when both inputs need broadcasting to
the common output shape, e.g. comparing shapes (2,2,4) and (1,4).

Use MultidirectionalBroadcastShape as in ROperator_BasicBinary, which is
covered by the new Comparison_broadcast_3d test.

guitargeek reviewed (approved):
LGTM! Thanks a lot for adding these tests, which also caught a real bug for which I added a fix to this PR.
